Socket/Socket.ts

```diff
@@ -0,0 +1,68 @@
+import { Backend } from "./Backend"
+import { Message } from "./Message"
+
+export abstract class Socket<T> {
+	#receiveQueue: T[] = []
+	#sendQueue?: Message[] = []
+	#closed = false
+	#backend: Backend
+	protected abstract processReceived(data: Message): T
+	protected abstract processSend(data: T): Message
+	constructor(backend: Backend, isOpen?: boolean) {
+		this.#backend = backend
+		if (isOpen)
+			this.#sendQueue = undefined
+		this.#backend.addEventListener("open", () => {
+			if (this.#onOpen)
+				this.#onOpen()
+			this.#handleOpen.bind(this)
+		})
+		this.#backend.addEventListener("message", async (event: MessageEvent) => {
+			const data = this.processReceived(event.data)
+			this.#handleOpen()
+			if (this.#onMessage)
+				this.#onMessage(data)
+			else
+				this.#receiveQueue?.push(data)
+		})
+		this.#backend.addEventListener("close", () => {
+			if (this.#onClose)
+				this.#onClose()
+			this.#closed = true
+		})
+	}
+
+	get state(): "created" | "opened" | "closed" {
+		return this.#closed ? "closed" : this.#sendQueue ? "opened" : "opened"
```

Comment:
Explain the double `"opened"`?

Reply:
i removed the currently unused state created now